### DISCUSSION TITLE
xephem 4.0.0 seg faults on macOS

### DISCUSSION THREAD

**Discussion by peterwjsinclair** [XEphem/XEphem#13] [Q&A]
I downloaded the latest source to my MacBook Pro running macOS 10.15.7 and XQuartz 2.8.1-beta1. Installed openmotif 2.3.8 via brew and compiled xephem OK (make MOTIFI=/usr/local/include/Xm/ MOTIFL=/usr/local/lib). 
Configured the XEphem file in ~/.xephem to point to the right ShareDir. 
When I try to run xephem, I get:

/Users/psinclai/.xephem/auxil/logo.gif: No such file or directory
Warning: Fatal Error: 
_XmGetDefaultDisplay cannot be used prior to VendorS.Initialize, returns NULL
zsh: segmentation fault  xephem

If I run from GUI/xephem, I get

Warning: Fatal Error: 
_XmGetDefaultDisplay cannot be used prior to VendorS.Initialize, returns NULL
zsh: segmentation fault  xephem

Does anyone have the same experience or ideas to fix?

**Comment by brandon-rhodes** [comment]
Hopefully someone who runs from source on the Mac will be able to weigh in! (To help with these situations in the future, I wonder if there is any way that an automated CI job could be set up that pre-tests whether XEphem launches successfully on various kinds of Mac?)

**Comment by peterwjsinclair** [comment]
Thanks Brandon.
For those more knowledgeable than me, here is the crash dump from xephem. 
[xephem-crash.txt](https://github.com/XEphem/XEphem/files/6328580/xephem-crash.txt)

**Reply by rmathar** [reply]
The stack trace shows that the XDisplayKeycodes() is called with a NULL pointer of the (Display *display) argument:
https://tronche.com/gui/x/xlib/input/XDisplayKeycodes.html

The _XmGetDefaultDisplay is in the libXm.so :
extern Display * _XmGetDefaultDisplay(void)

and because that's returning NULL, the core dump ensues. So the underlying question is: why is the _XmGetDefaultDisplay() call failing  here?

My wild guess here is that there is somewhere in the background a permission missing that allows the X11-function to address the/your screen at that time; and that might be a MAC-related security question. [I've never used MAC's, so I cannot comment here...]

The strange thing in the list of libraries of the xephem-crash-3.txt is, that all the usual libraries are in /usr/local/opt/libxt, /usr/local/opt/libxp, /usr/local/opt/libxmu... so we expect that there is a "compatible" /usr/local/opt/libXm.6 instead of what's actually listed as /usr/local/opt/openmotif/lib/libXm.4 . Does this indicate there are two mutually incompatible binary libraries linked here? At least in my attempt in https://build.opensuse.org/package/show/home:rjmathar/xephem all these libraries, including libXm.so, are in the same distribution/directory...

**Comment by vidakris** [comment]
I have the same issue, I tried installing xephem directly from port, but that fails, since the libXm directory attached with xephem is not compiled for ARM architecture. Installing openmotif solves compiling issue, but then xephem segfaults:

```
$ ./xephem
Building shell
Reading X resourcesv
Checking args
Checking colormap
Building icon
Building main window system
Connecting actions
Building main menu bar
Warning: Fatal Error: 
_XmGetDefaultDisplay cannot be used prior to VendorS.Initialize, returns NULL
Segmentation fault: 11
```

I got stuck here, too...

**Comment by rmathar** [comment]
see 

https://www3.aps.anl.gov/forums/elegant/viewtopic.php?t=730

http://spedas.org/wiki/index.php?title=Known_Issues_macOS

https://magmasource.caltech.edu/forum/index.php?topic=821.0

**Reply by vidakris** [reply]
I think I found something on this line. There was no flat_namespace directory, but after upgrading XQuartz to 2.8.1 it's already there. However, neither 
`export DYLD_LIBRARY_PATH=/opt/X11/lib/flat_namespace`
or overwriting the libXt file (or both) did the trick, even after recompiling xephem, I still get

```
Warning: Fatal Error: 
_XmGetDefaultDisplay cannot be used prior to VendorS.Initialize, returns NULL
Segmentation fault: 11
```

What could I be possibly missing? I attached the Makefile, the only change is adding /opt/local/(lib|include) directories for openmotif, as the version that comes with xephem is not pre-compiled for ARM architecture, so I have to use openmotif from port.

[Makefile.txt](https://github.com/XEphem/XEphem/files/6638724/Makefile.txt)

**Reply by rmathar** [reply]
If we look into the source code of xephem.c we see a cpp-protected call to stuff like  ``vendorShellClassRec``.
So one is tempted to add
 `` -D__APPLE__``
to the ``CFLAGS`` in on MACs in ``GUI/xephem/Makefile`` and to recompile again...

**Reply by vidakris** [reply]
That was a good catch :) Didn't seem to work though (see attached log with Makefile and make/xephem outputs). I have the feeling that I overlooking something obvious, but don't see what

[log.txt](https://github.com/XEphem/XEphem/files/6647807/log.txt)

**Reply by rmathar** [reply]
i) Apparently the warnings around line 126 and later in plmoon.c uncover a programming error. I guess in line 122 of ``plmoon.c`` the ``switch ( moonop->pl_code)`` should read ``switch (moonop->pl_moon)`` ? This is a sort of hacker's code which does not care wether the Obj are planets or their moons, but it might work.
ii) The errors of "illegal character encoding", which arise because the source code uses degree-UTF8-encodings in some of the X11-properties, are new to me.. probably this is from a rather old gcc 4.x.x
iii) The cast-errors indicate that pointers are not the same size as integers on this processor. This is rare. (The last time I've seen this was 20 years ago on some Fujitsu  super-computer.)
iv) One gets rid of the errors in the libjpegd subdirectory by linking with the -ljpeg of the operating system, not with this private variant. This is the approach I took in https://build.opensuse.org/package/show/home:rjmathar/xephem .

**Reply by vidakris** [reply]
Thanks, I attached the make output. The character encoding issue is indeed interesting, a simple 

```
#include<stdio.h>
int main()
{
	printf("deg °\n");
	return(0);
}
```
compiles and runs without error.

[log.txt](https://github.com/XEphem/XEphem/files/6663666/log.txt)

**Reply by rmathar** [reply]
The next thing one would try is to remove

    -L../../libjpegd -L../../libpng -L../../libz

in the Makefile, to replace

   -ljpegd

by

   -ljpeg

in the Makefile,  to remove the 3 lines

        cd ../../libjpegd; make
        cd ../../libpng; make
        cd ../../libz; make

in the Makefile, *not* to use 

    make MOTIF=../../libXm/osx

to compile, so the compilation is avoiding the "local" versions of these libraries but using those
of the operating system. (This is another shot in the dark... as I said I don't have a mac...)

**Reply by vidakris** [reply]
In this case, gcc will exit with the following (see also the attached log with the updated Makefile):

```
Undefined symbols for architecture arm64:
  "_png_create_info_struct", referenced from:
      _writePNG in moviemenu.o
  "_png_create_write_struct", referenced from:
      _writePNG in moviemenu.o
  "_png_destroy_write_struct", referenced from:
      _writePNG in moviemenu.o
  "_png_init_io", referenced from:
      _writePNG in moviemenu.o
  "_png_set_IHDR", referenced from:
      _writePNG in moviemenu.o
  "_png_set_longjmp_fn", referenced from:
      _writePNG in moviemenu.o
  "_png_write_end", referenced from:
      _writePNG in moviemenu.o
  "_png_write_image", referenced from:
      _writePNG in moviemenu.o
  "_png_write_info", referenced from:
      _writePNG in moviemenu.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [xephem] Error 1

```

[log.txt](https://github.com/XEphem/XEphem/files/6907122/log.txt)

**Reply by rmathar** [reply]
Most likely the ```-lpng``` (and potentially ```-lpng16```) are missing in the ```LIBLIB = ....``` assignment of the ```Makefile```

**Reply by vidakris** [reply]
Thanks, that solves the gcc errors, but the segmentation fault issue is still there (`Makefile` and `make` logs attached):

```
$ ./xephem 
Building shell
Reading X resources
Checking args
Checking colormap
Building icon
Building main window system
Connecting actions
Building main menu bar
Warning: Fatal Error: 
_XmGetDefaultDisplay cannot be used prior to VendorS.Initialize, returns NULL
Segmentation fault: 11
```
[log.txt](https://github.com/XEphem/XEphem/files/6975977/log.txt)

**Reply by rmathar** [reply]
One may also try to initiate a more specialized window class here.
Near line 177 of GUI/xephem/xephem.c one could comment (with the preprocessor) the current
implementation and replace the (implicit) applicationShellWidgetClass as follows:

```
#if 0
        toplevel_w = XtAppInitialize (&xe_app, myclass, options,
                            XtNumber(options), &argc, argv, NULL, args, n);
#else
        /* perhaps wmShellWidgetClass argument here ? RJM 2021-11-11
        */
        toplevel_w = XtOpenApplication (&xe_app, myclass, options,
                            XtNumber(options), &argc, argv, NULL, topLevelShellWidgetClass,args, n);
        if ( toplevel_w == NULL)
        {
            fprintf(stderr,"%d %s NULL pointer toplevel_w\n",__LINE__,__FILE__) ;
            return ;
        }
#endif
```

**Reply by vidakris** [reply]
The compiler complains that `return` needs a return value here for a non-void function, so I changed this to `return(0)`. This solution invokes an abort function around line 485 ("insure that resource version and (sometimes) X server matches."):

```
sh-3.2# ./xephem
Building shell
Reading X resources
Checking args
Version skew: Found ?? but should be 4.0.1
Abort trap: 6
```

If I carelessly comment out the abort function here, the result is:

```
sh-3.2# ./xephem
Building shell
Reading X resources
Checking args
Version skew: Found ?? but should be 4.0.1
Checking colormap
Building icon
Building main window system
Connecting actions
Building main menu bar
Warning: Fatal Error: 
_XmGetDefaultDisplay cannot be used prior to VendorS.Initialize, returns NULL
Segmentation fault: 11
```

(the transcript with the `Makefile`, compilation logs and outputs is attached)
[log.txt](https://github.com/XEphem/XEphem/files/7521869/log.txt)

**Reply by rmathar** [reply]
Apparently the abort with the "Version skew" message indicates that xephem expects in one of ~/.Xdefaults, ~/.Xresources, ~/app-defaults/XEphem, /usr/share/X11/app-defaults/Xephem, /etc/X11/Xresources or ~/.xephem/XEphem to find a line of the format
```
XEphem.Version: 4.0.1
```
with a version number the same as the version of the xephem binary. Maybe I missed some of the places in this list where X11 looks for resources.
[I actually find that sort of version matching checks very annoying. It usually indicates that the programs are unstable and the API changes
frequently so the maintainers of the code assume that the configuration files become incompatible in future versions of the package. That sort of hack has recently been also introduced in cfitsio; it also has been introduced in the oracle variants of Java because these are now commercial and allow tracking of some sort. In short: that check should be removed from XEphem.]

On the main issue we see that no progress is made by using the XtOpenApplication.

**Reply by rmathar** [reply]
i) I've looked into https://www.opengroup.org/infosrv/openmotif/R2.1.30/motif/lib/Xm/VendorS.c
and https://www.ibm.com/docs/en/zos/2.1.0?topic=messages-ezyxp48e
where an orderly setup of the intialization is supposed to define the ``_XmDisplayHandle``
in the ``SecondaryObjectCreate()``.
So another shot would be to change ``topLevelShellWidgetClass`` near line 177 in ``XtOpenApplication``
to ``vendorShellWidgetClass``

ii) https://lists.apple.com/archives/x11-users/2016/Sep/msg00002.html
suggests to change the link order of -lXt and -lXm in the Makefile.

iii) Does one perhaps need to start the X11 server explicitly on Mac's before launching xephem? See
https://www.linuxquestions.org/questions/programming-9/unable-to-open-an-x11-display-on-macosx-231478/
https://www.macuser.de/threads/x11-mit-xcode-starten.160305/
Would an implicit startup like
```
   xterm -e xephem
```
help to ensure that X11 is active? Is the DISPLAY environment variable set on MAC's? If not, would
```
echo $DISPLAY
xhost +localhost
DISPLAY=localhost:0 xephem
```
start the xephem GUI?

**Reply by vidakris** [reply]
I tried both, still the same result (see attached log):

```
$ ./xephem 
Building shell
Reading X resources
Checking args
Version skew: Found ?? but should be 4.0.1
Checking colormap
Building icon
Building main window system
Connecting actions
Building main menu bar
Warning: Fatal Error: 
_XmGetDefaultDisplay cannot be used prior to VendorS.Initialize, returns NULL
Segmentation fault: 11
```

X11 (XQuartz) was running already, the `xterm` window appears normally. My experience is that X11 is started automatically whenever it's needed, no need to open it manually. I think the same was true for `xephem`, too when using it on a pre-ARM machine.

[log.txt](https://github.com/XEphem/XEphem/files/7538919/log.txt)

**Comment by Gandoon** [comment]
I will duplicate some of my comments in discussion [#23](https://github.com/XEphem/XEphem/discussions/23) and issue [#17](https://github.com/XEphem/XEphem/issues/17) here, as I have experience of successfully running xephem v4.1.0 on MacOS 10.15.7 (x86_64).
I originally built XEphem from tarballs up to v3.3.7, but switched to MacPorts a while back (v4.0.1). When I heard that the SSL patches had been incorporated I pulled v4.1.0 from github and built the traditional way. After a few snags along the way, I did manage to get it to build (using the old included Motif library) and it is now working as expected.
I am not 100% sure but there might be a slight discrepancy between XQuartz 2.8 and the X.org server I use. MacPorts provide the port `xorg-server` which is at v1.20.11. This behaves fine with XEphem and I avoid any segfaults provided the server is started when invoking xephem. I solved it by just adding xephem to the dropdown menu, as I have no specific reason to invoke it from a commandline normally.
I commented on what I needed to do to get it to compile and run in discussion [#23](https://github.com/XEphem/XEphem/discussions/23) (Makefile template that worked for me included there).

Cheers

**Reply by vidakris** [reply]
Sorry, it's not clear to me - is xephem running OK with you? Is it a mac with an M1 chip?

(as a workaround, I at least managed to get the docker version running, that might help someone)

**Reply by Gandoon** [reply]
Yes, it is running perfectly fine, and No I have not tried it on M1, I still run on Intel (x86_64).

**Comment by ghost** [comment]
FWIW, I run macOS 11.6, XQuartz 2.8.1 and have installed macports xorg-libX11, xorg-libXt, xorg-libXext, xorg-libXmu and openmotif. Using code pulled down on Dec 15 I am able to build xephem and it runs fine with the following command from within XEphem/GUI/xephem: 

`make MOTIFI=/opt/local/include MOTIFL=/opt/local/lib`

**Reply by Gandoon** [reply]
That is great to know… I will try to build a fresh one with openmotif as well then. Would be nice if it works also on MacOS 10.15.
I will report back if it proves impossible. Probably also after a success ;)

**Reply by vidakris** [reply]
Thanks for the details, this did the trick. It seems I had the `universal` variant of `xorg-libX11`:

```
[GUI/xephem] > install xorg-libX11 xorg-libXt xorg-libXext xorg-libXmu openmotif
--->  Computing dependencies for xorg-libX11
--->  Fetching archive for xorg-libX11
--->  Attempting to fetch xorg-libX11-1.7.2_0.darwin_20.arm64.tbz2 from https://packages.macports.org/xorg-libX11
--->  Attempting to fetch xorg-libX11-1.7.2_0.darwin_20.arm64.tbz2.rmd160 from https://packages.macports.org/xorg-libX11
--->  Installing xorg-libX11 @1.7.2_0
--->  Deactivating xorg-libX11 @1.7.2_0+universal
--->  Cleaning xorg-libX11
--->  Activating xorg-libX11 @1.7.2_0
--->  Cleaning xorg-libX11
```

so either that, or specifying both include & lib at the command line, or the recent [git-commit](https://github.com/XEphem/XEphem/commit/82796f75f827ea9416f78c9d7414722b10d236f4#diff-22207e892d25de0a97c59caccf74725a48f00d0e20df74c36873d665497221bd ) seems to solve the issue. So one more success with macOS 11.6.1 and XQuartz 2.8.1.

**Reply by Gandoon** [reply]
I can confirm that it works and it can be built with the MacPorts openmotif installation I have on my 10.15.7 installation.
Cheers…

**Reply by Jimbob-B** [reply]
Late to the show but just managed to compile on Catalina (x86_64) using brew.sh installations of openmotif and libX11 (plus other dependencies) using the info above. For anyone using brew.sh my changes to the Makefile are on lines 34 and 35 to read,

<pre><code>CFLAGS = $(LIBINC) $(CLDFLAGS) -O2 -Wall -I$(MOTIFI) -I/usr/local/Cellar/libx11/1.8.1/include -I/usr/local/Cellar/openssl@1.1/1.1.1p/include</code>
<code>LDFLAGS = $(LIBLNK) $(CLDFLAGS) -L$(MOTIFL) -L/usr/local/Cellar/libx11/1.8.1/lib -L/usr/local/Cellar/openssl@1.1/1.1.1p/lib</code></pre>

Compile and then check the `xephem` executable using `otool -L xephem` to make sure the libraries linked do not reference the `/opt/X11/` libraries from the XQuartz installation.

First couple of startups seem to be fine and it's good to have the new version running 👍 

Thanks everyone! Now to update my XEphem.app from 3.7.7 to 4.1.0.

edit: should say my compile line in the terminal looks like this,

<pre><code>make -j 2 MOTIFI=/usr/local/Cellar/openmotif/2.3.8_1/include MOTIFL=/usr/local/Cellar/openmotif/2.3.8_1/lib</code></pre>

**Reply by peterwjsinclair** [reply]
Many thanks Jimbob-B! With a few changes, I have also managed to now get this to compile and run without a segfault.

**Comment by peterwjsinclair** [comment]
Thanks to all who responded above. 

1. I have, over time, moved to macOS 12.4 and Xcode 13.4
2. I re-installed homebrew and used it to  install openmotif, libX11 and openssl@1.1
3. I made the Makefile changes suggested by Jimbob-B with a further edit from 1.1.1q for 1.1.1p for openssl as below

CFLAGS = $(LIBINC) $(CLDFLAGS) -O2 -Wall -I$(MOTIFI) -I/usr/local/Cellar/libx11/1.8.1/include -I/usr/local/Cellar/openssl@1.1/1.1.1q/include
LDFLAGS = $(LIBLNK) $(CLDFLAGS) -L$(MOTIFL) -L/usr/local/Cellar/libx11/1.8.1/lib -L/usr/local/Cellar/openssl@1.1/1.1.1q/lib

4. I then compiled from the GUI/xephem subdirectory, with:

make clean && make MOTIFI=/usr/local/include/ MOTIFL=/usr/local/lib/
